lib/Runtime/Executor/ThreadPoolExecutor.cpp

```diff
@@ -0,0 +1,377 @@
+  // caller.
+  unsigned previousValue = inflightNodes_.fetch_sub(decrement);
+
+  // The decrement should never be more than the value of the counter at the
+  // time of decrement.
+  if (previousValue < decrement) {
+    assert(false && "More decrements than increments to inflight nodes!");
+  }
+
+  // Return true when the counter hits zero.
+  return (previousValue == decrement);
+}
+
+bool ExecutionState::incrementNodeParentsDone(const DAGNode *node,
+                                              unsigned increment) {
+  // Get the parents done counter for the node. It should have
+  // been created in the constructor.
+  auto it = nodeParentsDone_.find(node);
+
+  if (it == nodeParentsDone_.end()) {
+    assert(false && "Node parents done counter should exist but not found!");
+  }
+
+  // fetch_add must be used here so that the function returns true to only
+  // one caller.
+  unsigned numParents = (node->parents).size();
+  unsigned previousValue = (it->second).fetch_add(increment);
+  unsigned newValue = previousValue + increment;
+
+  // The new value of the counter cannot exceed the number of parents that
+  // the node has.
+  if (newValue > numParents) {
+    assert(false && "Node parents done counter incremented beyond limit!");
+  }
+
+  // Return true only when the counter hits the total numer of parents.
+  return (newValue == numParents);
+}
+
+void ExecutionState::insertIntoResultCtx(Placeholder *P, Tensor &&T) {
+  assert(resultCtx_ && "Execution result context should exist!");
+  std::lock_guard<std::mutex> lock(contextMtx_);
+  resultCtx_->insert(P, std::move(T));
+}
+
+std::unique_ptr<Context> ExecutionState::getUniqueResultContextPtr() {
+  assert(resultCtx_ && "Execution result context should exist!");
+  return std::move(resultCtx_);
+}
+
+void ExecutionState::setResultCode(const ResultCode resultCode) {
+  if (resultCode_ != ResultCode::Failed) {
+    resultCode_ = resultCode;
+  }
+}
+
+void ThreadPoolExecutor::run(const DAGNode *root,
+                             std::unique_ptr<Context> context,
+                             RunIdentifierTy runId, DoneCbTy cb) {
+  // If list of roots is empty, there is nothing to do. Give back the
+  // context so the caller can reuse it.
+  if (!root) {
+    cb(runId, ResultCode::Executed, std::move(context));
+    return;
+  }
+
+  // If the given run ID corresponds to a run already in progress, there is
+  // also nothing to do, but return an error. Give back the context so the
+  // caller can reuse it.
+  std::unique_lock<std::mutex> lock(executionStatesMutex_);
+  if (executionStates_.find(runId) != executionStates_.end()) {
+    cb(runId, ResultCode::Failed, std::move(context));
+    return;
+  }
+
+  // Create execution state tracker object for this run ID.
+  auto executionState =
+      std::make_shared<ExecutionState>(runId, root, std::move(cb));
+  executionStates_.insert(std::make_pair(runId, executionState));
+  lock.unlock();
```

Comment:
I really think it's safer to unlock via scope not a manual unlock() call, super easy to refactor this into a deadlock.

Reply:
I agree, but I did this because I don't think we should hold on to the lock as we call `propagatePlaceholdersForNode` and `executeDAGNode` below. Once the `runId` has been recorded in `executionStates_`, we don't need that lock anymore.

Reply:
What I'm suggesting is that you add another scope block when you grab the lock, and close it here where you would unlock - letting the lock going out of scope unlock it for you. Doesn't look like you create any variables that are referenced outside of this zone.

Reply:
Oh, that's a cool trick!